conans/client/toolchain/msbuild.py

```diff
@@ -0,0 +1,99 @@
+import os
+import textwrap
+from xml.dom import minidom
+
+from conans.client.tools import msvs_toolset
+from conans.util.files import save, load
+
+
+class MSBuildToolchain(object):
+
+    def __init__(self, conanfile):
+        self._conanfile = conanfile
+        self.definitions = {}
+
+    @ staticmethod
+    def _name_condition(settings):
+        toolset = msvs_toolset(settings)
+
+        props = [("Configuration", settings.build_type),
+                 # FIXME: This probably requires mapping ARM architectures
+                 ("Platform", {'x86': 'Win32',
+                               'x86_64': 'x64'}.get(settings.get_safe("arch"))),
+                 ("PlatformToolset", toolset)]
+
+        name = "".join("_%s" % v for _, v in props if v is not None)
+        condition = " And ".join("'$(%s)' == '%s'" % (k, v) for k, v in props if v is not None)
+        return name.lower(), condition
+
+    def write_toolchain_files(self):
+        name, condition = self._name_condition(self._conanfile.settings)
+        config_filename = "conan_toolchain{}.props".format(name)
+        self._write_config_toolchain(config_filename)
+        self._write_main_toolchain(config_filename, condition)
+
+    def _write_config_toolchain(self, config_filename):
+
+        def format_macro(k, value):
+            return '%s="%s"' % (k, value) if value is not None else k
+
+        runtime = self._conanfile.settings.get_safe("compiler.runtime")
+        cppstd = self._conanfile.settings.get_safe("compiler.cppstd")
+        runtime_library = {"MT": "MultiThreaded",
+                           "MTd": "MultiThreadedDebug",
+                           "MD": "MultiThreadedDLL",
+                           "MDd": "MultiThreadedDebugDLL"}.get(runtime, "")
+
+        content = textwrap.dedent("""\
+            <?xml version="1.0" encoding="utf-8"?>
+            <Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+              <ItemDefinitionGroup>
+                <ClCompile>
+                  <PreprocessorDefinitions>
+                     {};%(PreprocessorDefinitions)
+                  </PreprocessorDefinitions>
+                  <RuntimeLibrary>{}</RuntimeLibrary>
+                  <LanguageStandard>{}</LanguageStandard>
+                </ClCompile>
+              </ItemDefinitionGroup>
+            </Project>
+            """)
+        definitions = ";".join([format_macro(k, v) for k, v in self.definitions.items()])
```

Comment:
so, to be clear. for msbuild definitions, we want to set C++ preprocessor definitions like that:
```
self.msbuild.definitions["_NDEBUG"] = 1
```
but not msbuild properties (specified like `msbuild /p DeployOnBuild=True`):
```
self.msbuild.definitions["DeployOnBuild"] = True
```

I think it's kind fundamental, as it contradicts with current definitions concept of CMake toolchain.

Reply:
This is a very very good point, that the same function technically has a different meaning between in the  `CMakeToolchain`  and `MSBuildToolchain`. We should definitely reconcile this ASAP. 

Of note, in the `MakeToolchain`, I treated them as "preprocessor definitions" because that's what made the most sense to me at the time.  Now, I don't think it makes sense anymore.  I think the `CMakeToolchain` is the only one doing it right, and that we should update `MSBuild` and `Make` toolchains to implement definitions as "build system variables". 

If that's the direction we go, then the next question we should answer is this:

> Do we want to enable toolchain users to effectively declare pre-processor definitions?  If so, what would the syntax be since `definitions` is used to mean "build system variables".  Maybe "preprocessor_definitions"?

Whatever we do here, we should probably make sure we are consistent with build helpers too. 

Also of note, as a user, I had this problem in the past with `MSBuild` helper where I thought it meant "pre-processor definitions" and then learned the hard way that "it doesn't mean that".  Once I learned how it worked, it made sense to me. However, I did wish that I had a way to inject pre-processor definitions directly. But, just because I wanted that, doesn't mean lots of people will want it.  Just one user story.

Reply:
I agree with you @solvingj . in general, as toolchain feature is still in its early stages, and we can break it as it's experimental, I think we may change `Make` and `CMake` toolchain interfaces to be more explicit, e.g.:
```
self._msbuild.preprocessor_definitions["NDEBUG"] = 1
self._msbuild.build_definitions["DeployOnBuild"] = True
```
so there is no confusion. naming is rough here, maybe we can rename it to the properties/options/something else - I suppose we really need to chose non-obscure naming here to avoid confusion with prepocessor definitions, conan options, etc.

Reply:
Yes, I agree that naming on this one is difficult. 

After further consideration, because none of the build systems (cmake, make, msbuild) provide a dedicated command-line argument to set preprocessor definitions directly, so I don't think we should either. I'm sure the designers were all faced with the decision, and it has repeatedly been requested, and they've all still chosen not to implement it.

Reply:
Renamed to ``preprocessor_definitions``, the ``CMakeToolchain`` also contains this functionality (renamed in a separate PR). The build system ``properties`` would be a different feature.